core-bundle/src/EventListener/PreviewUrlCreateListener.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\ArticleModel;
+use Contao\CoreBundle\Event\PreviewUrlCreateEvent;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\PageModel;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * @internal
+ */
+class PreviewUrlCreateListener
+{
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var ContaoFramework
+     */
+    private $framework;
+
+    public function __construct(RequestStack $requestStack, ContaoFramework $framework)
+    {
+        $this->requestStack = $requestStack;
+        $this->framework = $framework;
+    }
+
+    /**
+     * Adds the page ID to the front end preview URL.
+     *
+     * @throws \RuntimeException
+     */
+    public function __invoke(PreviewUrlCreateEvent $event): void
+    {
+        if (!$this->framework->isInitialized() || !($id = $event->getId()) || !\in_array($event->getKey(), ['page', 'article'], true)) {
+            return;
+        }
+
+        $request = $this->requestStack->getCurrentRequest();
+
+        if (null === $request) {
+            throw new \RuntimeException('The request stack did not contain a request');
+        }
+
+        if ('article' === $event->getKey()) {
```

Comment:
I would move the condition here

```suggestion
        if (!$id = $event->getId()) {
            return;
        }

        if ('article' === $event->getKey()) {
```

and also replace

```php
        if (null === ($newsModel = $this->getNewsModel($this->getId($event, $request)))) {
            return;
        }
```

with

```php
        if ((!$id = $this->getId($event, $request)) || (!$newsModel = $this->getNewsModel($id))) {
            return;
        }
```

in the other `PreviewUrlCreateListener` classes. Then we are consistently handling the case that `$id` is empty.

Reply:
In my opinion, the ID should be checked before handling the request since nothing in the method will yield a result without an ID.

Maybe I am missing something and there is a reason to throw the `RuntimeException` if there is no request during an event with a key of `page` or `article ` but without an ID - instead of just ignoring the listener?

Reply:
> in the other `PreviewUrlCreateListener` classes. Then we are consistently handling the case that `$id` is empty.

I like this, saving a useless call to the databse if there is no ID. I will update the request.

Reply:
> In my opinion, the ID should be checked before handling the request since nothing in the method will yield a result without an ID.

You are right, checking the ID before retrieving the request makes more sense here. 👍